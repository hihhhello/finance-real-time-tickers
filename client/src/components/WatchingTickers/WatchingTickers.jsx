import React from "react";
import _ from "lodash";
import Ticker from "../Ticker";

import styles from "./watching-tickers.module.scss";
import { useActions } from "../../hooks/useActions";
import { useSelector } from "react-redux";
import { SkeletonTicker } from "../Ticker/SkeletonTicker";

const TickersSkeleton = () => {
  const placeholder = [];
  for (let i = 0; i < 6; i++) {
    placeholder.push(<Ticker key={i} skeleton />);
  }
  return placeholder;
};

export const WatchingTickers = () => {
  const { deleteTicker } = useActions();
  const { tickers } = useSelector((state) => state.tickers);
  const { socket } = useSelector((state) => state.socket);
  const userSettings = useSelector((state) => state.userSettings);

  const onDeleteTicker = (ticker) => {
    deleteTicker({ ticker, socket });
  };
  return (
    <div className={styles.wrapper}>
      {_.isNull(tickers) && userSettings && userSettings.isWatching ? (
        <TickersSkeleton />
      ) : userSettings.watchingTickers &&
        userSettings.watchingTickers.length &&
        !userSettings.isWatching &&
        _.isNull(tickers) ? (
        userSettings.watchingTickers.map((t) => (
          <SkeletonTicker key={t} ticker={t} />
        ))
      ) : _.isEmpty(tickers) ? (
        <div className={styles.placeholder}>
          <span>No watching tickers</span>
        </div>
      ) : (
        Object.entries(tickers).map(([ticker, [lastTickerData]]) => (
          <Ticker
            key={ticker}
            ticker={ticker}
            {...lastTickerData}
            onClickHandler={onDeleteTicker}
          />
        ))
      )}
    </div>
  );
};
