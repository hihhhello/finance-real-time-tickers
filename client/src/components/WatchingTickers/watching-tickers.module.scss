@import "../../scss/var";
.wrapper {
  font-size: 20rem;
  padding-top: 0.175em;
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(1em, 1fr));
  grid-auto-rows: auto;
  justify-content: center;
  grid-gap: 1rem;
}

.placeholder {
  display: grid;
  place-items: center;
  color: $gray-color;
  font-size: 0.125em;
  border: 2px solid $gray-color;
  min-height: 3em;
  width: 100%;
  grid-column: 1 / -1;
  grid-row: 1 / span 2;
  border-radius: 0.5em;
}
