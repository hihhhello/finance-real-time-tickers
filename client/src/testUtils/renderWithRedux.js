import { Provider } from "react-redux";
import { combineReducers, createStore } from "redux";
import { render } from "@testing-library/react";
import reducers from "../store/slices";

export const renderWithRedux = (
  component,
  {
    initialState,
    store = createStore(combineReducers(reducers), initialState),
  } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};
