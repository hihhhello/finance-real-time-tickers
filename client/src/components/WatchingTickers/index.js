import { WatchingTickers } from "./WatchingTickers";
export default WatchingTickers;
