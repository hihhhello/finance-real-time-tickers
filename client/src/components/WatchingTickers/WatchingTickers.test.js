import WatchingTickers from ".";
import { renderWithRedux } from "../../testUtils/renderWithRedux";

const initialState = {
  socket: {
    socket: {},
  },
  tickers: {
    tickers: null,
  },
};

describe("<WatchingTickers />", () => {
  it("loading", () => {
    const { asFragment } = renderWithRedux(<WatchingTickers />);

    expect(asFragment(<WatchingTickers />)).toMatchSnapshot();
  });

  it("renders w/o watching tickers", () => {
    const { asFragment, getByText } = renderWithRedux(<WatchingTickers />, {
      initialState: {
        ...initialState,
        tickers: {
          tickers: {},
        },
      },
    });
    expect(getByText(/No watching tickers/)).toBeInTheDocument();
    expect(asFragment(<WatchingTickers />)).toMatchSnapshot();
  });

  it("renders with tickers", () => {
    const tickers = {
      TSLA: [
        {
          exchange: "NASDAQ",
          price: "276.43",
          change: "-105.18",
          changePercent: "-0.90",
          dividend: "0.11",
          yield: "1.23",
          lastTradeTime: "2021-09-28T12:47:55.000Z",
        },
      ],
      GOOGL: [
        {
          exchange: "NASDAQ",
          price: "173.29",
          change: "-52.05",
          changePercent: "-0.67",
          dividend: "0.86",
          yield: "1.70",
          lastTradeTime: "2021-09-28T12:48:25.000Z",
        },
      ],
    };

    const { asFragment, getByText } = renderWithRedux(<WatchingTickers />, {
      initialState: {
        ...initialState,
        tickers: {
          tickers: {
            ...tickers,
          },
        },
      },
    });
    expect(getByText("TSLA")).toBeInTheDocument();
    expect(getByText("GOOGL")).toBeInTheDocument();
    expect(asFragment(<WatchingTickers />)).toMatchSnapshot();
  });
});
