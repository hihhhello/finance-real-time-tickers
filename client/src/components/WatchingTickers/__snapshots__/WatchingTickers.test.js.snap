// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WatchingTickers /> loading 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="placeholder"
    >
      <span>
        No watching tickers
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<WatchingTickers /> renders w/o watching tickers 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="placeholder"
    >
      <span>
        No watching tickers
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<WatchingTickers /> renders with tickers 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="wrapper wrapperLoss wrapperPaused"
    >
      <div
        class="tickWrapper tickWrapperLoss"
      >
        <svg
          class="tick"
        >
          arrow-down-solid.svg
        </svg>
      </div>
      <div
        class="textWrapper"
      >
        <div
          class="name"
        >
          TSLA
        </div>
        <div
          class="priceWrapper"
        >
          <div
            class="price"
          >
            276,43
          </div>
        </div>
      </div>
      <div
        class="changeWrapper"
      >
        <div
          class="changePercentWrapper"
        >
          <div
            class="changePercent"
          >
            -0.90
          </div>
        </div>
        <div
          class="changePriceWrapper"
        >
          <div
            class="change"
          >
            -105.18
          </div>
        </div>
      </div>
    </div>
    <div
      class="wrapper wrapperLoss wrapperPaused"
    >
      <div
        class="tickWrapper tickWrapperLoss"
      >
        <svg
          class="tick"
        >
          arrow-down-solid.svg
        </svg>
      </div>
      <div
        class="textWrapper"
      >
        <div
          class="name"
        >
          GOOGL
        </div>
        <div
          class="priceWrapper"
        >
          <div
            class="price"
          >
            173,29
          </div>
        </div>
      </div>
      <div
        class="changeWrapper"
      >
        <div
          class="changePercentWrapper"
        >
          <div
            class="changePercent"
          >
            -0.67
          </div>
        </div>
        <div
          class="changePriceWrapper"
        >
          <div
            class="change"
          >
            -52.05
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
